sw/device/silicon_creator/lib/base/sec_mmio_unittest.cc

```diff
@@ -0,0 +1,189 @@
+  EXPECT_EQ(GetShutDownCount(), 0);
+  EXPECT_EQ(ctx_->last_index, 2);
+}
+
+TEST_F(SecMmioTest, Write32SimulatedFault) {
+  EXPECT_ABS_WRITE32(mmio_, 0, 0x12345678);
+  EXPECT_ABS_READ32(mmio_, 0, 0);
+  sec_mmio_write32(0, 0x12345678);
+  EXPECT_EQ(GetShutDownCount(), 1);
+}
+
+TEST_F(SecMmioTest, CounterInc) {
+  sec_mmio_write_increment(5);
+  EXPECT_EQ(ctx_->expected_write_count, 5);
+
+  sec_mmio_write_increment(10);
+  EXPECT_EQ(ctx_->expected_write_count, 15);
+}
+
+TEST_F(SecMmioTest, CheckValues) {
+  EXPECT_ABS_WRITE32(mmio_, 0, 0x12345678);
+  EXPECT_ABS_READ32(mmio_, 0, 0x12345678);
+  sec_mmio_write32(0, 0x12345678);
+
+  EXPECT_ABS_WRITE32(mmio_, 4, 0x87654321);
+  EXPECT_ABS_READ32(mmio_, 4, 0x87654321);
+  sec_mmio_write32(4, 0x87654321);
+
+  EXPECT_ABS_WRITE32(mmio_, 8, 0);
+  EXPECT_ABS_READ32(mmio_, 8, 0);
+  sec_mmio_write32(8, 0);
+
+  // The expected permutation order for rnd_offset=0 is {1, 2, 0}.
+  EXPECT_ABS_READ32(mmio_, 4, 0x87654321);
+  EXPECT_ABS_READ32(mmio_, 8, 0);
+  EXPECT_ABS_READ32(mmio_, 0, 0x12345678);
+  sec_mmio_check_values(/*rnd_offset=*/0);
+  EXPECT_EQ(ctx_->check_count, 1);
+
+  // The expected permutation order for rnd_offset=1 is {2, 0, 1}.
+  EXPECT_ABS_READ32(mmio_, 8, 0);
+  EXPECT_ABS_READ32(mmio_, 0, 0x12345678);
+  EXPECT_ABS_READ32(mmio_, 4, 0x87654321);
+  sec_mmio_check_values(/*rnd_offset=*/1);
+  EXPECT_EQ(ctx_->check_count, 2);
+
+  // The expected permutation order for rnd_offset=32 is {0, 1, 2}.
+  // rnd_offset=32 value is not relevant.
+  EXPECT_ABS_READ32(mmio_, 0, 0x12345678);
+  EXPECT_ABS_READ32(mmio_, 4, 0x87654321);
+  EXPECT_ABS_READ32(mmio_, 8, 0);
+  sec_mmio_check_values(/*rnd_offset=*/32);
+  EXPECT_EQ(ctx_->check_count, 3);
+
+  EXPECT_EQ(GetShutDownCount(), 0);
+}
+
+TEST_F(SecMmioTest, CheckValuesSimlatedFault) {
+  EXPECT_ABS_WRITE32(mmio_, 0, 0x12345678);
+  EXPECT_ABS_READ32(mmio_, 0, 0x12345678);
+  sec_mmio_write32(0, 0x12345678);
+
+  EXPECT_ABS_READ32(mmio_, 0, 0);
+  sec_mmio_check_values(/*rnd_offset=*/0);
+  EXPECT_EQ(GetShutDownCount(), 1);
+}
+
+TEST_F(SecMmioTest, CheckCount) {
+  EXPECT_ABS_WRITE32(mmio_, 0, 0x12345678);
+  EXPECT_ABS_READ32(mmio_, 0, 0x12345678);
+  sec_mmio_write32(0, 0x12345678);
+  sec_mmio_write_increment(1);
+
+  sec_mmio_check_counters(/*expected_check_count=*/0);
+  sec_mmio_check_counters(/*expected_check_count=*/1);
+  EXPECT_EQ(ctx_->check_count, 2);
+  EXPECT_EQ(GetShutDownCount(), 0);
+}
+
+TEST_F(SecMmioTest, CheckCountWriteMismatch) {
+  EXPECT_ABS_WRITE32(mmio_, 0, 0x12345678);
+  EXPECT_ABS_READ32(mmio_, 0, 0x12345678);
+  sec_mmio_write32(0, 0x12345678);
+
+  // The developer forgot to increment the write counter, or an attacker
+  // glitched the sec write operation.
+  sec_mmio_check_counters(/*expected_check_count=*/0);
+
+  // The count is incremented twice per sec_mmio_check_counters call
+  // on any failure condition due to redundant result checks.
+  EXPECT_EQ(GetShutDownCount(), 2);
```

Comment:
This seems to test undefined behavior because the control flow reaches __builtin_unreachable, right?

Reply:
Yeah, that should compile down to a `ud2` so I don't think this test should pass at all.

I'm in favor of wrapping `__builtin_unreachable` such that it calls abort() when we're running tests.

Reply:
I need to look at this to understand why the test is passing.

Reply:
switched to use `std::abort()` and `ASSERT_DEATH`.